File: backend/app/events/event_bus.py
from collections.abc import Awaitable, Callable
from typing import Any


EventHandler = Callable[[dict[str, Any]], Awaitable[None]]


class EventBus:
    """Placeholder async event bus for future service decoupling."""

    def __init__(self) -> None:
        self._handlers: dict[str, list[EventHandler]] = {}

    def subscribe(self, event_type: str, handler: EventHandler) -> None:
        self._handlers.setdefault(event_type, []).append(handler)

    async def publish(self, event_type: str, payload: dict[str, Any]) -> None:
        for handler in self._handlers.get(event_type, []):
            await handler(payload)


event_bus = EventBus()

File: frontend/lib/events/event-bus.ts
export type AppEventType =
  | "QuestCompleted"
  | "RecipeCrafted"
  | "EmotionRestored"
  | "WeatherChanged"
  | "CompanionReacted"
  | "SettingsUpdated";

export type AppEvent<TPayload = unknown> = {
  type: AppEventType;
  payload: TPayload;
};

type EventHandler<TPayload = unknown> = (event: AppEvent<TPayload>) => void;

class EventBus {
  private handlers = new Map<AppEventType, Set<EventHandler>>();

  subscribe<TPayload>(type: AppEventType, handler: EventHandler<TPayload>) {
    const handlers = this.handlers.get(type) ?? new Set<EventHandler>();
    handlers.add(handler as EventHandler);
    this.handlers.set(type, handlers);

    return () => {
      handlers.delete(handler as EventHandler);
    };
  }

  publish<TPayload>(event: AppEvent<TPayload>) {
    this.handlers.get(event.type)?.forEach((handler) => handler(event));
  }
}

export const eventBus = new EventBus();
